planner/core/expression_rewriter.go

```diff
@@ -1280,12 +1280,12 @@ func (er *expressionRewriter) rewriteVariable(v *ast.VariableExpr) {
 			// Normally we can infer the type from SessionVars.User, but we need SessionVars.UserVarTypes when
 			// GetVar has not been executed to fill the SessionVars.Users.
 			sessionVars.UsersLock.Lock()
```

Comment:
Please move this lock operation into the new interface function `Set/GetUserVar`.